test/test_images/probe_helper/probe_helper_test.go

```diff
@@ -14,67 +14,343 @@ limitations under the License.
+	return &event
+}
+
+func testPubsubClient(ctx context.Context, t *testing.T, projectID string) (*pubsub.Client, func()) {
+	srv := pstest.NewServer()
+	conn, err := grpc.Dial(srv.Addr, grpc.WithInsecure())
+	if err != nil {
+		t.Fatalf("Failed to dial test pubsub connection: %v", err)
+	}
+	close := func() {
+		srv.Close()
+		conn.Close()
+	}
+	c, err := pubsub.NewClient(ctx, projectID, option.WithGRPCConn(conn))
+	if err != nil {
+		t.Fatalf("Failed to create test pubsub client: %v", err)
+	}
+	return c, close
+}
+
+func testStorageClient(ctx context.Context, t *testing.T) (*storage.Client, chan *http.Request, func()) {
+	gotRequest := make(chan *http.Request, 1)
+	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// The test Cloud Storage server forwards the client's generated HTTP requests.
+		body, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			logging.FromContext(ctx).Fatal("Test Cloud Storage server could not read request body.")
+		}
+		r.Body = ioutil.NopCloser(bytes.NewBuffer(body))
+		gotRequest <- r
+		w.Write([]byte("{}"))
+	}))
+	c, err := storage.NewClient(ctx, option.WithEndpoint(srv.URL))
+	if err != nil {
+		t.Fatalf("Failed to create test storage client: %v", err)
+	}
+	return c, gotRequest, srv.Close
+}
+
+type eventAndResult struct {
+	event      *cloudevents.Event
+	wantResult protocol.Result
+}
+
+func TestProbeHelper(t *testing.T) {
+	os.Setenv("K_SINK", testBrokerURL)
+	os.Setenv("PROJECT_ID", testProjectID)
+	os.Setenv("CLOUDPUBSUBSOURCE_TOPIC_ID", testTopicID)
+	os.Setenv("CLOUDSTORAGESOURCE_BUCKET_ID", testStorageBucket)
+
+	cases := []struct {
+		name  string
+		steps []eventAndResult
+	}{{
+		name: "Broker E2E delivery probe",
+		steps: []eventAndResult{
+			{
+				event:      probeEvent("broker-e2e-delivery-probe", ""),
+				wantResult: cloudevents.ResultACK,
+			},
+		},
+	}, {
+		name: "CloudPubSubSource probe",
+		steps: []eventAndResult{
+			{
+				event:      probeEvent("cloudpubsubsource-probe", ""),
+				wantResult: cloudevents.ResultACK,
+			},
+		},
+	}, {
+		name: "CloudStorageSource probe",
+		steps: []eventAndResult{
+			{
+				event:      probeEvent("cloudstoragesource-probe", "create"),
+				wantResult: cloudevents.ResultACK,
+			},
+			{
+				event:      probeEvent("cloudstoragesource-probe", "update-metadata"),
+				wantResult: cloudevents.ResultACK,
+			},
+			{
+				event:      probeEvent("cloudstoragesource-probe", "archive"),
+				wantResult: cloudevents.ResultACK,
+			},
+			{
+				event:      probeEvent("cloudstoragesource-probe", "delete"),
+				wantResult: cloudevents.ResultACK,
+			},
+		},
+	}, {
+		name: "Unrecognized probe event type",
+		steps: []eventAndResult{
+			{
+				event:      probeEvent("unrecognized-probe-type", ""),
+				wantResult: cloudevents.ResultNACK,
+			},
+		},
+	}}
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			ctx := logtest.TestContextWithLogger(t)
+			ctx = WithProjectKey(ctx, testProjectID)
```

Comment:
Where is WithTopicKey defined? I couldn't find it

Reply:
https://github.com/google/knative-gcp/blob/master/pkg/pubsub/adapter/context/topic.go#L27

Reply:
Thanks.

I don't think these are useful since we only have one topic in the tests. Same for other context attributes.

Reply:
I needed these in order to get the converter in the fake CloudPubSubSource to work: https://github.com/google/knative-gcp/blob/master/pkg/pubsub/adapter/converters/pubsub.go

I can instead manually craft the CloudEvent, if you think that is better?

Reply:
OK I see. Never mind then.